### ISSUE TITLE
Adding 2D flexible mesh conversion

### ISSUE THREAD

**Issue by bekozi** [pyugrid/pyugrid#96]
Hi, all. I've talked with @ChrisBarker-NOAA about bringing in some code we have for 2D flexible mesh conversion to/from UGRID. I listed below the main changes required to bring in this code (see https://github.com/NESII/ocgis-ugrid/tree/spatial-index-speedups for the most recent version of our flexible mesh conversion code). Some refactoring of our original code will be required, but I don't expect major changes in pyugrid.

* Add a new base class object mirroring ``Ugrid`` called ``AbstractUgridObject``. A new ``FlexibleMesh`` object will inherit from this. ``Ugrid`` can also be made a child class. Universal methods such as ``write_nc_dataset`` will be abstract methods on the new base class.
* Add a package called ``flexible_mesh`` to the ``pyugrid`` package containing new code.
* Some additional dependencies will be required for reading from / conversion to shapefile: mpi4py, rtree, shapely, and fiona. The imports for these libraries exist only in the flexible mesh code (i.e. the original triangular code will still function without them). Mpi4py and rtree can be made completely optional.

I hope this all looks okay. Let me know if you have any suggestions. Looking forward to working with you!

**Comment by ChrisBarker-NOAA** [comment]
Welcome!

1st: what is a "flexible mesh" -- i.e. how much can change from time to time:
 - number of cells?
 - topology -- i.e neighbors.

I'm trying to get  a sense of if you need entire new ugrid (essentially) for each time (or each time the mesh changes...) -- and how is data, etc mapped between meshes as the mesh changes?

Sorry for the maybe obvious questions -- never worked with one....

And as for base class, we are working, inthe APIRUS project on a prototype for ABC -- maybe this will derive from that same ABC -- and we can use mix-ins to get the shared functionality.

Also, are you only concerned with file format conversion for the moment?

**Comment by bekozi** [comment]
> 1st: what is a "flexible mesh" -- i.e. how much can change from time to time:
number of cells?
topology -- i.e neighbors.
I'm trying to get a sense of if you need entire new ugrid (essentially) for each time (or each time the mesh changes...) -- and how is data, etc mapped between meshes as the mesh changes?

The specification is here: https://github.com/ugrid-conventions/ugrid-conventions/blob/v0.9.0/ugrid-conventions.md#2d-flexible-mesh-mixed-triangles-quadrilaterals-etc-topology. There is no time-varying component to the mesh per se. It is "flexible" as individual faces may vary in topology and node counts. For example, one can represent U.S. state boundaries as a flexible mesh. To make the mesh time varying, I expect one would want to use separate files or extend the convention somehow.

> Sorry for the maybe obvious questions -- never worked with one....

Nothing about flexible meshes was obvious to me when I started working with them :-). The name is a bit misleading.

> And as for base class, we are working, inthe APIRUS project on a prototype for ABC -- maybe this will derive from that same ABC -- and we can use mix-ins to get the shared functionality.

Sure, whatever you think is best. To simplify things in that case, it makes sense to have the new ``FlexibleMesh`` object inherit from ``Ugrid`` and not fuss with a new base class immediately.

> Also, are you only concerned with file format conversion for the moment?

Not necessarily. Our use case for the flexible mesh conversion has been extracting hydrologic catchments from shapefile to UGRID for use in ESMF mesh regridding. What did you have in mind?

**Comment by rsignell-usgs** [comment]
@bekozi, just to make sure I understand, you are persisting shapefile-like objects in NetCDF files using UGRID conventions?

**Comment by bekozi** [comment]
@rsignell-usgs: That is correct. The conversion works for arbitrary collections of "polygon objects" (shapely polygons to be exact). Also, we hope to propose some additions to UGRID for handling MultiPolygons eventually. Implementing ragged arrays for node indexing will be crucial at some point. Currently, flexible mesh UGRID NetCDFs are larger on disk than their shapefile counterparts!  Two other missing elements for full polygon shapefile interoperability are:
 1. Only the unique identifier is maintained from the shapefile. Other attribute fields are stripped.
 2. Supporting CF and arbitrary coordinate systems. Lat/lon geographic is assumed.

**Comment by rsignell-usgs** [comment]
@bekozi, Interesing.  I would have guessed that with netcdf4 deflation the file sizes of the netcdf file and shapefile have been similar. 
Could you point to an example of one of these netcdf files?  I would be very curious to check out one.

**Comment by bekozi** [comment]
@rsignell-usgs: An example is here: https://www.dropbox.com/sh/dizuyn2wj8y39ot/AAB63Kie7Nj7Vu_lvZx87s3xa?dl=0.

I initially thought the file sizes would be the same as well. There are some unfortunately shaped variables in flexible mesh with dimension (nMesh2_face, nMaxMesh2_face_nodes) corresponding to (the number of faces/polgyons, the maximum number of nodes contained in a face/polygon). The kicker is that you may be converting a file with 5 polygons - four of the polygons have 5 nodes while the fifth has 1000. The shape of the indexing variables will be (5, 1000) even though 99% of that array is masked. Ragged arrays would address this issue and bring netCDF storage in line with shapefile. This is mentioned in the spec...

Note that no compression is used on the example netCDF!

**Comment by rsignell-usgs** [comment]
@bekozi , I'm happy to report that simply converting this netCDF3 file to netCDF4 with deflation resulted in a file slightly *smaller* than the zipfile, with no loss of information.  A win!
```
[testbed@testbed2 ugrid]$ ls -lh
total 1.8M
-rw-rw-r--. 1 testbed testbed 161K Sep 22 16:59 state_boundaries_shapefile.zip
-rw-rw-r--. 1 testbed testbed 1.6M Sep 22 16:59 ugrid_state_boundaries.nc
```
and using the handy `nc3tonc4` python script included with `netcdf4-python`:
```
[testbed@testbed2 ugrid]$ nc3tonc4 ugrid_state_boundaries.nc ugrid_state_boundaries_nc4.nc
copying global attributes ..
copying dimensions ..
copying variable Mesh2
copying variable Mesh2_face_x
copying variable Mesh2_face_y
copying variable Mesh2_face_ids
copying variable Mesh2_face_nodes
copying variable Mesh2_edge_nodes
copying variable Mesh2_face_edges
copying variable Mesh2_face_links
copying variable Mesh2_node_x
copying variable Mesh2_node_y

[testbed@testbed2 ugrid]$ ls -lh *
-rw-rw-r--. 1 testbed testbed 161K Sep 22 16:59 state_boundaries_shapefile.zip
-rw-rw-r--. 1 testbed testbed 1.6M Sep 22 16:59 ugrid_state_boundaries.nc
-rw-rw-r--. 1 testbed testbed 159K Sep 22 17:09 ugrid_state_boundaries_nc4.nc
```

**Comment by ChrisBarker-NOAA** [comment]
"""
four of the polygons have 5 nodes while the fifth has 1000.
"""

Ahh -- was about to ask that question. So while netcdf4 compression should help the file size issue, for that kin dof "mesh" -- this simply isn't the way to go.

I think that spec was designed for things like quad+triangle meshes, where there would only be a little bit of wasted space. What we have here is really a bunch of polygons. So the ragged array approach makes much more sense.

For that matter is this really a mesh at all? will things like face_face_connectivity make sense?

I'm sure I've seen some discussion f how to represent more GIS-y data in netcdf -- not sure if it came to anything, but this seems to be more about "how to represent the GIS data models in netcdf", and if that can share some with UGRID, great, but it's not really a UGRID anymore...

Googling for CF and netcdf and polygons, I found a nice thread on the CF list -- guess who was on that thread?

Ben and myself, of course -- so what goes around comes around :-)

NOTE: I use ragged arrays for storing output from particle tracking models -- in a "standard" format that was born of much discussion but I never published anywhere (still intend to). But you can find some Python code for workign with it here:

https://github.com/NOAA-ORR-ERD/GnomeTools/blob/master/post_gnome/post_gnome/nc_particles.py

-CHB

**Comment by rsignell-usgs** [comment]
@ChrisBarker-NOAA, it seems to me this *is* a perfectly fine way to represent this GIS data in NetCDF files: it's simple and the netcdf4 file sizes are comparable to shapefiles.    As I understand it, data from these files will not be used to create UGRID objects in python, but instead used to instantiate objects in Shapely and the like.   

The only drawback I can see is that the memory requirements will be temporarily bigger during a read or write than if a ragged array was used, but until that's documented to be real problem, this seems a fine approach.

**Comment by ChrisBarker-NOAA** [comment]
Maybe I'm just being pedantic, but storing a huge number of empty values in order to squish something into a simple data model feel like a big kludge. Even if compression can save you from big files.

Maybe that temporarily bigger issues isn't that big -- we'll see.

But ragged arrays really aren't THAT hard to deal with :-)

Anyway, this may be the pragamatic way to go -- and having py_ugrid support the general case is a good thing.

**Comment by bekozi** [comment]
> I'm happy to report that simply converting this netCDF3 file to netCDF4 with deflation resulted in a file slightly smaller than the zipfile, with no loss of information. A win!

@rsignell-usgs: This is excellent! We had not experimented with file compression. Thanks for testing!

> For that matter is this really a mesh at all? will things like face_face_connectivity make sense?

@ChrisBarker-NOAA: They make sense provided the input file has connectivity. Isolated polygons will have no links while things like catchments should, provided they are not on an island. The conversion does a "touches" operation to determine face connectivity.

> Googling for CF and netcdf and polygons, I found a nice thread on the CF list -- guess who was on that thread?

@ChrisBarker-NOAA: I remember that one :-). I have successfully stored polygons and multi-polygons using ragged arrays in netCDF. So yeah, totally doable.

> As I understand it, data from these files will not be used to create UGRID objects in python, but instead used to instantiate objects in Shapely and the like.

@rsignell-usgs: That is pretty much correct. Clients that understand flexible mesh UGRID can use the data (likely the UGRID file) directly (e.g. ESMF). Python UGRID objects would still be created with Shapely as an output. Flexible mesh is very similar to triangular mesh with only the dimension sizes changing to account for varying node counts. The files also have nifty features like a partially computed spatial index (when face bounds are added) and built in topology in the presence of connectivity.

> The only drawback I can see is that the memory requirements will be temporarily bigger during a read or write than if a ragged array was used, but until that's documented to be real problem, this seems a fine approach.

@rsignell-usgs: We'll have to look more into compression as a solution for the large node index arrays. The shapefiles we're using create index arrays with dimensions like (81000, 18000). These sizes are generally intractable unless one is very careful. For many use cases, these node counts will be uncommon and file compression can address most issues I expect.

> Anyway, this may be the pragmatic way to go -- and having py_ugrid support the general case is a good thing.

@ChrisBarker-NOAA: I agree. A keyword argument like "use_ragged_arrays" could be added to the netCDF file conversion methods.

Are you still leaning towards tight integration with ``FlexibleMesh`` inheriting from ``Ugrid``?

**Comment by ChrisBarker-NOAA** [comment]
Yes, I think FlexibleMesh could be tightly integrated with UGRID -- in fact, I'm not sure there needs to be anything separate at all -- particularly if you don't go the ragged array route, the only thing you need to add is handling of missing values.

At first thought, it seems that could be totally transparent to the user...

-Chris

**Comment by bekozi** [comment]
Sounds good! I'll get back to you with successes and/or failures. Ragged arrays will not be supported initially.

**Comment by ChrisBarker-NOAA** [comment]
cool, thanks!

**Comment by bekozi** [comment]
I'd like to reboot the pull request for the flexible mesh implementation. The main holdup with the last attempt was the addition of multiple dependencies. I propose to make the flexible mesh object independent from the IO calls (i.e. read from shapefile, write to shapefile, write to ESMF format). This will keep the installation cleaner. Import errors will be thrown for the IO calls that require things like ``fiona``, etc.

I have two other proposed changes that effect the current code:
 1. The current ``UGrid`` object has a number of ``build_*`` methods that I believe are specific to 2d triangular meshes. I would like to refactor this class slightly to extract an ``AbstractUGridObject`` class that has the base methods - no build-like methods or other fields that are specific to triangles. The ``UGrid`` object name for triangular meshes would remain intact.
 2. I'd like to add support for keyword arguments to ``netcdf4.Dataset`` creation in ``save_as_netcdf``. Mostly for using the netCDF3-classic model. Eventually, I would like to get ragged array support, but I expect that should go in when the UGRID convention supports it.
 
Are there major ``pyugrid`` changes incoming? If not, I could start with the base class refactor and netCDF save changes. Thanks!

**Comment by ChrisBarker-NOAA** [comment]
@bekozi: are these changes included in #142 ? or are yougiving up, or???

It did look like good stuff -- sorry we never worked out the kinks -- I'd rather it didn't all get lost.
